honeybadgermpc/progs/fixedpoint.py

```diff
@@ -0,0 +1,229 @@
+
+
+async def bit_ltl(ctx, a, b_bits):
+    """
+    a: Public
+    b: List of private bit shares. Least significant digit first
+    """
+    b_bits = [ctx.Share(Field(1) - bi.v) for bi in b_bits]
+    a_bits = [ctx.Share(ai) for ai in binary_repr(int(a), len(b_bits))]
+
+    # a_bits_opened = [(await a_bit.open()) for a_bit in a_bits]
+    # logging.debug("a: ", a_bits_opened)
+
+    # b_bits_opened = [(await b_bit.open()) for b_bit in b_bits]
+    # logging.debug("b: ", b_bits_opened)
+
+    carry = await get_carry_bit(ctx, a_bits, b_bits)
+    return ctx.Share(Field(1) - carry.v)
+
+
+async def mod2m(ctx, x, k, m):
+    r1, r1_bits = await random2m(ctx, m)
+    r2, _ = await random2m(ctx, k + KAPPA - m)
+    r2 = ctx.Share(r2.v * Field(2) ** m)
+
+    c = await (x + r2 + r1 + Field(2) ** (k - 1)).open()
+    c2 = int(c) % (2 ** m)
+    u = await bit_ltl(ctx, c2, r1_bits)
+    a2 = ctx.Share(Field(c2) - r1.v + (2 ** m) * u.v)
+    return a2
+
+
+async def trunc(ctx, x, k, m):
+    a2 = await mod2m(ctx, x, k, m)
+    d = ctx.Share((x.v - a2.v) / (Field(2)) ** m)
+    return d
+
+
+class FixedPoint(object):
+    def __init__(self, ctx, x):
+        self.ctx = ctx
+        if type(x) in [int, float]:
+            self.share = ctx.preproc.get_zero(ctx) + ctx.Share(int(x * 2 ** F))
+        elif type(x) is ctx.Share:
+            self.share = x
+        else:
+            raise NotImplementedError
+
+    def add(self, x):
+        if type(x) is FixedPoint:
+            return FixedPoint(self.ctx, self.share + x.share)
+
+    def sub(self, x):
+        if type(x) is FixedPoint:
+            return FixedPoint(self.ctx, self.share - x.share)
+        raise NotImplementedError
+
+    async def mul(self, x):
+        if type(x) is FixedPoint:
+            start_time = time.time()
+            res_share = await (self.share * x.share)
+            end_time = time.time()
+            logging.info("Multiplication time: %.2f", end_time - start_time)
+            start_time = time.time()
+            res_share = await trunc_pr(self.ctx, res_share, 2 * K, F)
+            end_time = time.time()
+            logging.info("Trunc time: %.2f", end_time - start_time)
+            return FixedPoint(self.ctx, res_share)
+        raise NotImplementedError
+
+    async def open(self):
+        x = (await self.share.open()).value
+        if x >= 2 ** (K - 1):
+            x = -(p - x)
+        return float(x) / 2 ** F
+
+    def neg(self):
+        return FixedPoint(self.ctx, Field(-1) * self.share)
+
+    async def ltz(self):
+        t = await trunc(self.ctx, self.share, K, K - 1)
+        return self.ctx.Share(-t.v)
+
+    async def lt(self, x):
+        return await self.sub(x).ltz()
+
+    async def div(self, x):
+        if type(x) in [float, int]:
+            return await self.mul(FixedPoint(self.ctx, 1. / x))
+        raise NotImplementedError
+
+
+async def _prog(ctx):
+    ctx.preproc = FakePreProcessedElements()
```

Comment:
I don't think this is a good idea-- won't this file not otherwise work with other programs unless they also modify context like this? At the least we should document this.